components/script/dom/imagedata.rs

```diff
@@ -47,122 +51,92 @@ impl ImageData {
+            pixelFormat: ImageDataPixelFormat::Rgba_unorm8,
         };
 
-        let byte_len = unsafe { typed_array.as_slice().len() } as u32;
-        if byte_len == 0 || byte_len % 4 != 0 {
-            return Err(Error::InvalidState);
-        }
+        if let Some(ref mut d) = data {
+            d.resize(len as usize, 0);
 
-        let len = byte_len / 4;
-        if width == 0 || len % width != 0 {
-            return Err(Error::IndexSize);
-        }
+            let cx = GlobalScope::get_cx();
+            rooted!(in (*cx) let mut js_object = std::ptr::null_mut::<JSObject>());
+            auto_root!(in(*cx) let data = create_buffer_source::<ClampedU8>(cx, &d[..], js_object.handle_mut(), can_gc)
+                        .map_err(|_| Error::JSFailed)?);
 
-        let height = len / width;
-        if opt_height.is_some_and(|x| height != x) {
-            return Err(Error::IndexSize);
+            Self::Constructor_(global, None, can_gc, data, width, Some(height), &settings)
+        } else {
+            Self::Constructor(global, None, can_gc, width, height, &settings)
         }
-
-        let imagedata = Box::new(ImageData {
-            reflector_: Reflector::new(),
-            width,
-            height,
-            data: heap_typed_array,
-        });
-
-        Ok(reflect_dom_object_with_proto(
-            imagedata, global, proto, can_gc,
-        ))
     }
 
-    fn new_without_data(
+    #[allow(clippy::too_many_arguments)]
+    /// <https://html.spec.whatwg.org/multipage/#initialize-an-imagedata-object>
+    fn initialize(
+        pixels_per_row: u32,
+        rows: u32,
+        settings: &ImageDataSettings,
+        source: Option<CustomAutoRooterGuard<Uint8ClampedArray>>,
+        default_color_space: Option<PredefinedColorSpace>,
         global: &GlobalScope,
         proto: Option<HandleObject>,
-        width: u32,
-        height: u32,
         can_gc: CanGc,
     ) -> Fallible<DomRoot<ImageData>> {
-        // If one or both of sw and sh are zero, then throw an "IndexSizeError" DOMException.
-        if width == 0 || height == 0 {
-            return Err(Error::IndexSize);
-        }
-
-        // When a constructor is called for an ImageData that is too large, other browsers throw
-        // IndexSizeError rather than RangeError here, so we do the same.
-        let len =
-            pixels::compute_rgba8_byte_length_if_within_limit(width as usize, height as usize)
-                .ok_or(Error::IndexSize)?;
-
-        let cx = GlobalScope::get_cx();
-
-        let heap_typed_array =
-            create_heap_buffer_source_with_length::<ClampedU8>(cx, len as u32, can_gc)?;
-
-        let imagedata = Box::new(ImageData {
-            reflector_: Reflector::new(),
-            width,
-            height,
-            data: heap_typed_array,
-        });
+        // 1. If source was given:
+        let data = if let Some(source) = source {
+            HeapBufferSource::<ClampedU8>::from_view(source)
```

Comment:
This seems to be missing the first two checks, and you may also want to add a doc for "// 1.3 Initialize the data attribute of imageData to source"